pageserver/src/walingest.rs

```diff
@@ -400,39 +398,72 @@ impl WalIngest {
                 // the size of tuple data is inferred from the size of the record.
                 // we can't validate the remaining number of bytes without parsing
                 // the tuple data.
-                if (xlrec.flags & pg_constants::XLH_UPDATE_NEW_ALL_VISIBLE_CLEARED) != 0 {
-                    timeline.put_wal_record(
-                        lsn,
-                        RelishTag::Relation(RelTag {
-                            forknum: pg_constants::VISIBILITYMAP_FORKNUM,
-                            spcnode: decoded.blocks[0].rnode_spcnode,
-                            dbnode: decoded.blocks[0].rnode_dbnode,
-                            relnode: decoded.blocks[0].rnode_relnode,
-                        }),
-                        decoded.blocks[0].blkno / pg_constants::HEAPBLOCKS_PER_PAGE as u32,
-                        ZenithWalRecord::ClearVisibilityMapFlags {
-                            heap_blkno: decoded.blocks[0].blkno,
-                            flags: pg_constants::VISIBILITYMAP_VALID_BITS,
-                        },
-                    )?;
-                }
-                if (xlrec.flags & pg_constants::XLH_UPDATE_OLD_ALL_VISIBLE_CLEARED) != 0
-                    && decoded.blocks.len() > 1
+                if (xlrec.flags & pg_constants::XLH_UPDATE_NEW_ALL_VISIBLE_CLEARED) != 0
+                    || ((xlrec.flags & pg_constants::XLH_UPDATE_OLD_ALL_VISIBLE_CLEARED) != 0
+                        && decoded.blocks.len() > 1)
                 {
-                    timeline.put_wal_record(
-                        lsn,
-                        RelishTag::Relation(RelTag {
-                            forknum: pg_constants::VISIBILITYMAP_FORKNUM,
-                            spcnode: decoded.blocks[1].rnode_spcnode,
-                            dbnode: decoded.blocks[1].rnode_dbnode,
-                            relnode: decoded.blocks[1].rnode_relnode,
-                        }),
-                        decoded.blocks[1].blkno / pg_constants::HEAPBLOCKS_PER_PAGE as u32,
-                        ZenithWalRecord::ClearVisibilityMapFlags {
-                            heap_blkno: decoded.blocks[1].blkno,
-                            flags: pg_constants::VISIBILITYMAP_VALID_BITS,
-                        },
-                    )?;
+                    if (xlrec.flags
```

Comment:
It took me a while to understand the logic here.

Looking at XLH_UPDATE_NEW_ALL_VISIBLE_CLEARED should only be set, if decoded.block.len() > 1. When the old and the new tuple are on the same page, PostgreSQL only sets XLH_UPDATE_OLD_ALL_VISIBLE_CLEARED. I was confused why that case is not handled here. A comments, and perhaps a warning if it does happen after all, would be in order.

But this doesn't seem to actually handle the case that only XLH_UPDATE_OLD_ALL_VISIBLE_CLEARED is set, and decoded.blocks.len() == 1.

We really need a regression test for this, to check that the result on-disk is the same with this and with vanilla PostgreSQL WAL replay.

Reply:
Maybe you could use `match` here to cover all the different combinations in a more readable format

Reply:
> It took me a while to understand the logic here.
> 
> Looking at XLH_UPDATE_NEW_ALL_VISIBLE_CLEARED should only be set, if decoded.block.len() > 1. When the old and the new tuple are on the same page, PostgreSQL only sets XLH_UPDATE_OLD_ALL_VISIBLE_CLEARED. I was confused why that case is not handled here. A comments, and perhaps a warning if it does happen after all, would be in order.
> 
> But this doesn't seem to actually handle the case that only XLH_UPDATE_OLD_ALL_VISIBLE_CLEARED is set, and decoded.blocks.len() == 1.
> 
> We really need a regression test for this, to check that the result on-disk is the same with this and with vanilla PostgreSQL WAL replay.

I handle two cases:
1. When UPDATE record contains both OLD/NEW VISIBLE_CLEARED  and them are mapped to the same VM page (but records can be located at different heap pages).
I expect that in this case decoded.block.len() is always 2, but I left this check because it was present in original code. Only in this case ClearVisibilityMapFlags record updating two VM locations is generated.
2. Otherwise both bits are checked independently as it was done before. The code looks pretty ugly, but logic is quite simple...